qucumber/rbm/purification_rbm.py

```diff
@@ -0,0 +1,533 @@
+        super().__init__()
+        self.num_visible = int(num_visible)
+        self.num_hidden = int(num_hidden)
+        self.num_aux = int(num_aux)
+
+        # Parameters are:
+        # W: The weights of the visible-hidden edges
+        # U: The weights of the visible-auxiliary edges
+        # b: The biases of the visible nodes
+        # c: The biases of the hidden nobdes
+        # d: The biases of the auxiliary nodes
+
+        # The auxiliary bias of the phase RBM is always zero
+
+        self.num_pars = (
+            (self.num_visible * self.num_hidden)
+            + (self.num_visible * self.num_aux)
+            + self.num_visible
+            + self.num_hidden
+            + self.num_aux
+        )
+
+        _warn_on_missing_gpu(gpu)
+        self.gpu = gpu and torch.cuda.is_available()
+
+        self.device = torch.device("cuda") if self.gpu else torch.device("cpu")
+
+        self.initialize_parameters(zero_init=zero_init)
+
+    def initialize_parameters(self, zero_init=False):
+        """Initialize the parameters of the RBM
+        :param zero_init: Whether or not to initalize the weights to zero
+        :type zero_init: bool
+        """
+        if zero_init:
+            gen_tensor = torch.zeros
+        else:
+            gen_tensor = torch.rand
+
+        self.weights_W = nn.Parameter(
+            (
+                0.01 * gen_tensor(self.num_hidden, self.num_visible, dtype=torch.double)
+                - 0.005
+            ),
+            requires_grad=False,
+        )
+
+        self.weights_U = nn.Parameter(
+            (
+                0.01 * gen_tensor(self.num_aux, self.num_visible, dtype=torch.double)
+                - 0.005
+            ),
+            requires_grad=False,
+        )
+
+        self.visible_bias = nn.Parameter(
+            torch.zeros(self.num_visible, dtype=torch.double), requires_grad=False
+        )
+
+        self.hidden_bias = nn.Parameter(
+            torch.zeros(self.num_hidden, dtype=torch.double), requires_grad=False
+        )
+
+        self.aux_bias = nn.Parameter(
+            torch.zeros(self.num_aux, dtype=torch.double), requires_grad=False
+        )
+
+    def effective_energy(self, v, a):
+        """Computes the equivalent of the "effective energy" for the RBM
+        :param v: The current state of the visible units
+        :type v: torch.Tensor
+        :param a: The current state of the auxiliary units
+        :type v: torch.Tensor
+        :returns: The "effective energy" of the RBM
+        :rtype: torch.Tensor
+        """
+        if len(v.shape) < 2 and len(a.shape) < 2:
+            v = v.unsqueeze(0)
+            a = a.unsqueeze(0)
+
+            energy = torch.zeros(v.shape[0], a.shape[0], dtype=torch.double)
+
+            vb_term = torch.mv(v, self.visible_bias)
+            ab_term = torch.mv(a, self.aux_bias)
+            vb_term += torch.mv(torch.matmul(v, self.weights_U.data.t()), self.aux_bias)
```

Comment:
Excellent catch, I've changed that. Although I don't think this is called in training though so it shouldn't affect that.

Reply:
i believe it would affect the computed value of the partition function which would mean the fidelity calculations and other exact metrics are off by a little bit. This might also explain why the gradient testing is a bit unpredictable.

Reply:
Partition function computed by trace of constructed matrix. Still this change has been made if the partition function is every required